src/v/serde/serde.h

```diff
@@ -0,0 +1,314 @@
+
+template<typename T>
+struct help_has_serde_async_read<
+  T,
+  std::void_t<decltype(std::declval<T>().serde_async_read(
+    std::declval<std::add_lvalue_reference_t<iobuf_parser>>(),
+    version_t{0},
+    version_t{0},
+    size_t{0U}))>> : std::true_type {};
+
+template<typename T>
+inline constexpr auto const has_serde_async_read
+  = help_has_serde_async_read<T>::value;
+
+template<typename T, typename = void>
+struct help_has_serde_async_write : std::false_type {};
+
+template<typename T>
+struct help_has_serde_async_write<
+  T,
+  std::void_t<decltype(std::declval<T>().serde_async_write(
+    std::declval<std::add_lvalue_reference_t<iobuf>>()))>> : std::true_type {};
+
+template<typename T>
+inline constexpr auto const has_serde_async_write
+  = help_has_serde_async_write<T>::value;
+
+template<typename T>
+concept HasAsyncWrite = requires(T a) {
+    {
+        a.serde_async_write(std::declval<std::add_lvalue_reference_t<iobuf>>())
+        } -> std::convertible_to<ss::future<>>;
+};
+
+template<typename T>
+inline constexpr auto const is_serde_compatible_v
+  = is_envelope_v<T> || (std::is_scalar_v<T> && !std::is_enum_v<T>)
+    || reflection::is_std_vector_v<
+      T> || reflection::is_named_type_v<T> || reflection::is_ss_bool_v<T> || std::is_same_v<T, std::chrono::milliseconds> || std::is_same_v<T, iobuf> || std::is_same_v<T, ss::sstring> || reflection::is_std_optional_v<T>;
+
+using header_t = std::tuple<version_t, version_t, size_t>;
+
+#if defined(SERDE_TEST)
+using serde_size_t = uint16_t;
+#else
+using serde_size_t = int32_t;
+#endif
+
+template<typename T>
+void write(iobuf& out, T const& t);
+
+template<HasAsyncWrite T>
+void write_envelope(iobuf& out, T const& t) {
+    t.serde_write(out);
+}
+
+template<typename T>
+void write_envelope(iobuf& out, T const& t) {
+    envelope_for_each_field(t, [&out](auto& f) { write(out, f); });
+}
+
+template<typename T>
+void write(iobuf& out, T const& t) {
+    using Type = std::decay_t<T>;
+    static_assert(has_serde_write<Type> || is_serde_compatible_v<Type>);
+
+    if constexpr (is_envelope_v<Type>) {
+        write(out, Type::redpanda_serde_version);
+        write(out, Type::redpanda_serde_compat_version);
+
+        auto size_placeholder = out.reserve(sizeof(serde_size_t));
+        auto const size_before = out.size_bytes();
+
+        write_envelope(out, t);
+
+        auto const written_size = out.size_bytes() - size_before;
+        if (unlikely(written_size > std::numeric_limits<serde_size_t>::max())) {
+            throw serde_exception("envelope too big");
+        }
+        auto const size = ss::cpu_to_le(
+          static_cast<serde_size_t>(written_size));
+        size_placeholder.write(
+          reinterpret_cast<char const*>(&size), sizeof(serde_size_t));
+    } else if constexpr (std::is_same_v<bool, Type>) {
+        write<int8_t>(out, t);
+    } else if constexpr (std::is_scalar_v<Type> && !std::is_enum_v<Type>) {
```

Comment:
maybe just std::is_arithmetic since scalar handles stuff like pointers that we don't care about

Reply:
double check float/double wire encoding

Reply:
Scalars work all the same:
https://github.com/google/flatbuffers/blob/6ed780dbd50f3fd4f2a12b13f5ef709da5372401/include/flatbuffers/base.h#L390-L396